File: modernized/bill-payment/src/services/billPaymentService.ts
import { Repositories } from '../repositories';
import { Account, isValidAcctId, normalizeAcctId } from '../models/account';
import {
  BILL_PAYMENT_TXN,
  Transaction,
  buildTimestamp,
  formatTransactionId,
} from '../models/transaction';
import { BillPaymentError, BILL_PAYMENT_MESSAGES } from '../errors';
import { BillPaymentRequest, BillPaymentResponse } from '../dto/billPayment';

/**
 * Bill Payment service — the modern equivalent of the COBIL00C
 * PROCESS-ENTER-KEY paragraph and its helper paragraphs
 * (READ-ACCTDAT-FILE, READ-CXACAIX-FILE, STARTBR/READPREV/WRITE-TRANSACT-FILE,
 * UPDATE-ACCTDAT-FILE).
 *
 * Business rule: a bill payment pays the account balance in full. It records a
 * transaction for the full current balance and then sets the balance to zero.
 */
export class BillPaymentService {
  constructor(
    private readonly repos: Repositories,
    private readonly clock: () => Date = () => new Date(),
  ) {}

  /**
   * Look up an account and return its current balance without paying anything.
   * Mirrors the COBIL00C path where CONFIRM is blank: READ-ACCTDAT-FILE only.
   */
  async getBalance(rawAcctId: string): Promise<BillPaymentResponse> {
    const account = await this.loadAccount(rawAcctId);
    return {
      paid: false,
      acctId: account.acctId,
      currentBalance: account.currentBalance.toString(),
      message: '',
    };
  }

  /**
   * Process the Enter key from the Bill Payment screen.
   *
   * @throws BillPaymentError for every validated failure path (empty/invalid
   *   account id, bad confirm value, account not found, nothing to pay, ...),
   *   matching the COBOL error paragraphs.
   */
  async processEnter(request: BillPaymentRequest): Promise<BillPaymentResponse> {
    // WHEN ACTIDINI = SPACES OR LOW-VALUES -> 'Acct ID can NOT be empty...'
    const rawAcctId = (request.acctId ?? '').trim();
    if (rawAcctId === '') {
      throw new BillPaymentError('ACCT_ID_REQUIRED');
    }

    // EVALUATE CONFIRMI: classify the confirm flag (case-insensitive).
    const confirm = (request.confirm ?? '').trim().toUpperCase();
    if (confirm !== '' && confirm !== 'Y' && confirm !== 'N') {
      // WHEN OTHER -> 'Invalid value. Valid values are (Y/N)...'
      throw new BillPaymentError('INVALID_CONFIRM');
    }
    if (confirm === 'N') {
      // WHEN 'N'/'n' -> CLEAR-CURRENT-SCREEN, set err flag (abort, nothing paid).
      return {
        paid: false,
        acctId: normalizeAcctId(rawAcctId),
        currentBalance: '',
        message: '',
      };
    }

    // READ-ACCTDAT-FILE (UPDATE intent — we will rewrite it if we pay).
    const account = await this.loadAccount(rawAcctId);

    // IF ACCT-CURR-BAL <= ZEROS -> 'You have nothing to pay...'
    if (account.currentBalance.isZeroOrLess()) {
      throw new BillPaymentError('NOTHING_TO_PAY');
    }

    // Blank confirm: display balance and prompt, do not pay.
    if (confirm !== 'Y') {
      return {
        paid: false,
        acctId: account.acctId,
        currentBalance: account.currentBalance.toString(),
        message: 'Confirm to make a bill payment...',
      };
    }

    // CONF-PAY-YES branch — execute the payment.
    return this.executePayment(account);
  }

  /** READ-ACCTDAT-FILE with NOTFND handling. */
  private async loadAccount(rawAcctId: string): Promise<Account> {
    const acctId = normalizeAcctId(rawAcctId);
    if (!isValidAcctId(acctId)) {
      throw new BillPaymentError('ACCT_ID_INVALID');
    }
    const account = await this.repos.accounts.findById(acctId);
    if (!account) {
      throw new BillPaymentError('ACCOUNT_NOT_FOUND');
    }
    return account;
  }

  private async executePayment(account: Account): Promise<BillPaymentResponse> {
    // READ-CXACAIX-FILE — resolve the card number for the account.
    const xref = await this.repos.cardXrefs.findByAcctId(account.acctId);
    if (!xref) {
      throw new BillPaymentError('XREF_NOT_FOUND');
    }

    // STARTBR(HIGH-VALUES) + READPREV + ENDBR -> highest tran id; ADD 1.
    const nextId = (await this.repos.transactions.maxTransactionId()) + 1;
    const amount = account.currentBalance; // pay balance in full

    const transaction: Transaction = {
      transactionId: formatTransactionId(nextId),
      typeCode: BILL_PAYMENT_TXN.TYPE_CODE,
      categoryCode: BILL_PAYMENT_TXN.CATEGORY_CODE,
      source: BILL_PAYMENT_TXN.SOURCE,
      description: BILL_PAYMENT_TXN.DESCRIPTION,
      amount,
      merchantId: BILL_PAYMENT_TXN.MERCHANT_ID,
      merchantName: BILL_PAYMENT_TXN.MERCHANT_NAME,
      merchantCity: BILL_PAYMENT_TXN.MERCHANT_CITY,
      merchantZip: BILL_PAYMENT_TXN.MERCHANT_ZIP,
      cardNumber: xref.cardNumber,
      originTimestamp: buildTimestamp(this.clock()),
      processTimestamp: buildTimestamp(this.clock()),
    };

    // WRITE-TRANSACT-FILE
    await this.repos.transactions.insert(transaction);

    // COMPUTE ACCT-CURR-BAL = ACCT-CURR-BAL - TRAN-AMT ; UPDATE-ACCTDAT-FILE
    const updated: Account = {
      ...account,
      currentBalance: account.currentBalance.subtract(amount),
    };
    await this.repos.accounts.update(updated);

    return {
      paid: true,
      acctId: updated.acctId,
      currentBalance: updated.currentBalance.toString(),
      message: `Payment successful. Your Transaction ID is ${transaction.transactionId}.`,
      transactionId: transaction.transactionId,
    };
  }
}

export { BILL_PAYMENT_MESSAGES };

File: java/src/main/java/com/carddemo/online/billpay/BillPaymentService.java
package com.carddemo.online.billpay;

import com.carddemo.model.entity.Account;
import com.carddemo.model.entity.CardXref;
import com.carddemo.model.entity.Transaction;
import com.carddemo.online.transaction.TransactionBrowseRepository;
import com.carddemo.online.transaction.TransactionFormats;
import com.carddemo.repository.AccountRepository;
import com.carddemo.repository.CardXrefRepository;
import com.carddemo.repository.TransactionRepository;
import com.carddemo.util.CobolUtils;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Clock;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Optional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * COBOL program: COBIL00C — "Pay bill of the account in full" (transaction CB00).
 *
 * <p>Files: ACCTDAT (copybook CVACT01Y), CXACAIX (copybook CVACT03Y), TRANSACT (copybook
 * CVTRA05Y). Screen: BMS mapset COBIL00, map COBIL0A.
 *
 * <p>A confirmed payment writes one transaction for the whole current balance (type code 02,
 * category 2, source POS TERM, merchant 999999999 "BILL PAYMENT") and then rewrites the account
 * with {@code ACCT-CURR-BAL = ACCT-CURR-BAL - TRAN-AMT}. COBIL00C touches no other account field:
 * the cycle credit and debit totals are deliberately left alone, as the batch posting programs own
 * them.
 */
@Service
public class BillPaymentService {

    static final String MSG_ACCT_EMPTY = "Acct ID can NOT be empty...";
    static final String MSG_CONFIRM_INVALID = "Invalid value. Valid values are (Y/N)...";
    static final String MSG_ACCT_NOT_FOUND = "Account ID NOT found...";
    static final String MSG_NOTHING_TO_PAY = "You have nothing to pay...";
    static final String MSG_CONFIRM = "Confirm to make a bill payment...";
    static final String MSG_TRAN_ID_EXISTS = "Tran ID already exist...";

    /** TRAN-TYPE-CD moved by PROCESS-ENTER-KEY. */
    static final String PAYMENT_TYPE_CODE = "02";

    private static final DateTimeFormatter TIMESTAMP =
            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    private final AccountRepository accounts;
    private final CardXrefRepository cardXrefs;
    private final TransactionRepository transactions;
    private final TransactionBrowseRepository transactionBrowse;
    private final Clock clock;

    @Autowired
    public BillPaymentService(
            AccountRepository accounts,
            CardXrefRepository cardXrefs,
            TransactionRepository transactions,
            TransactionBrowseRepository transactionBrowse) {
        this(accounts, cardXrefs, transactions, transactionBrowse, Clock.systemDefaultZone());
    }

    /** EXEC CICS ASKTIME / FORMATTIME is driven by an injectable clock so tests stay deterministic. */
    BillPaymentService(
            AccountRepository accounts,
            CardXrefRepository cardXrefs,
            TransactionRepository transactions,
            TransactionBrowseRepository transactionBrowse,
            Clock clock) {
        this.accounts = accounts;
        this.cardXrefs = cardXrefs;
        this.transactions = transactions;
        this.transactionBrowse = transactionBrowse;
        this.clock = clock;
    }

    /** PROCESS-ENTER-KEY. */
    @Transactional
    public BillPaymentResponse pay(BillPaymentRequest request) {
        String accountIdInput = request.getAccountId() == null ? "" : request.getAccountId().trim();
        if (accountIdInput.isEmpty()) {
            return failure(MSG_ACCT_EMPTY, null);
        }

        String confirm = request.getConfirm() == null ? "" : request.getConfirm().trim();
        boolean confirmed;
        if (confirm.equalsIgnoreCase("Y")) {
            confirmed = true;
        } else if (confirm.equalsIgnoreCase("N")) {
            // CLEAR-CURRENT-SCREEN then MOVE 'Y' TO WS-ERR-FLG: the screen is wiped, no message.
            return failure(null, null);
        } else if (confirm.isEmpty()) {
            confirmed = false;
        } else {
            return failure(MSG_CONFIRM_INVALID, null);
        }

        // ACTIDIN is moved to ACCT-ID PIC 9(11); a value that is not a number can never match a key.
        Optional<Account> found = isNumeric(accountIdInput)
                ? accounts.findById(Long.parseLong(accountIdInput))
                : Optional.empty();
        if (found.isEmpty()) {
            return failure(MSG_ACCT_NOT_FOUND, null);
        }
        Account account = found.get();
        BigDecimal balance = CobolUtils.nvl(account.getCurrentBalance());

        if (balance.signum() <= 0) {
            return failure(MSG_NOTHING_TO_PAY, balance);
        }

        if (!confirmed) {
            return failure(MSG_CONFIRM, balance);
        }

        CardXref xref = cardXrefs.findFirstByAccountId(account.getAccountId()).orElse(null);
        if (xref == null) {
            return failure(MSG_ACCT_NOT_FOUND, balance);
        }

        String transactionId = nextTransactionId();
        if (transactions.existsById(transactionId)) {
            return failure(MSG_TRAN_ID_EXISTS, balance);
        }

        // MOVE ACCT-CURR-BAL TO TRAN-AMT: the whole outstanding balance is paid in one transaction.
        BigDecimal paidAmount = balance;
        String timestamp = LocalDateTime.now(clock).format(TIMESTAMP) + ".000000";
        transactions.save(Transaction.builder()
                .transactionId(transactionId)
                .typeCode(PAYMENT_TYPE_CODE)
                .categoryCode(2)
                .source("POS TERM")
                .description("BILL PAYMENT - ONLINE")
                .amount(paidAmount)
                .cardNumber(xref.getCardNumber())
                .merchantId(999999999L)
                .merchantName("BILL PAYMENT")
                .merchantCity("N/A")
                .merchantZip("N/A")
                .originTimestamp(timestamp)
                .processTimestamp(timestamp)
                .build());

        BigDecimal newBalance = balance.subtract(paidAmount);
        account.setCurrentBalance(newBalance);
        accounts.save(account);

        return BillPaymentResponse.builder()
                .success(true)
                .message("Payment successful.  Your Transaction ID is " + transactionId + ".")
                .currentBalanceDisplay(currentBalance(newBalance))
                .currentBalance(newBalance)
                .transactionId(transactionId)
                .paidAmount(paidAmount)
                .build();
    }

    private static boolean isNumeric(String value) {
        for (int i = 0; i < value.length(); i++) {
            if (!Character.isDigit(value.charAt(i))) {
                return false;
            }
        }
        return !value.isEmpty();
    }

    /** STARTBR on HIGH-VALUES + READPREV + 1; an empty file leaves TRAN-ID at zeros. */
    private String nextTransactionId() {
        long highest = transactionBrowse
                .findFirstByOrderByTransactionIdDesc()
                .map(transaction -> Long.parseLong(transaction.getTransactionId().trim()))
                .orElse(0L);
        return TransactionFormats.transactionId(highest + 1);
    }

    private BillPaymentResponse failure(String message, BigDecimal balance) {
        return BillPaymentResponse.builder()
                .success(false)
                .message(message)
                .currentBalance(balance)
                .currentBalanceDisplay(balance == null ? null : currentBalance(balance))
                .build();
    }

    /** Renders WS-CURR-BAL PIC +9999999999.99. */
    static String currentBalance(BigDecimal value) {
        BigDecimal scaled = CobolUtils.nvl(value).setScale(2, RoundingMode.HALF_UP);
        String digits = scaled.abs().unscaledValue().toString();
        if (digits.length() < 3) {
            digits = "0".repeat(3 - digits.length()) + digits;
        }
        String integerPart = digits.substring(0, digits.length() - 2);
        String decimalPart = digits.substring(digits.length() - 2);
        if (integerPart.length() < 10) {
            integerPart = "0".repeat(10 - integerPart.length()) + integerPart;
        } else if (integerPart.length() > 10) {
            integerPart = integerPart.substring(integerPart.length() - 10);
        }
        return (scaled.signum() < 0 ? "-" : "+") + integerPart + "." + decimalPart;
    }
}
